pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,244 @@
+
+            # create a tuple of sampler names
+            samplers_names = tuple(v.__class__.__name__ for k, v in dataset_dict.items() if isinstance(v, Sampler))
+
+            # create a dictionary of generator present within the dataset attributes
+            dataset_sampler_generators = {k: v for k, v in dataset_dict.items() if isinstance(v, Generator)}
+
+            # iterate over the generator. If a generator was created from a ``Sampler```,
+            # it will be wrapped into a ``FastForwardSampler``.
+            for (generator_attr_name, generator) in dataset_sampler_generators.items():
+
+                # Generator name have the  the form `SamplerName.__iter__`
+                generator_name = generator.__qualname__.split('.')[0]
+
+                # validate the base generator name matches a sampler name.
+                if any(sampler_name == generator_name for sampler_name in samplers_names):
+
+                    # wrap the generator into a ``FastForwardSampler``
+                    sampler = FastForwardSampler(generator)
+
+                    # if ``CaptureIterableDataset`` was available, the sampler should reload its own state.
+                    if self.state_dict is not None:
+                        sampler.load_state_dict(self.state_dict[generator_attr_name])
+
+                    # store the samplers
+                    self.samplers[generator_attr_name] = sampler
+
+                    # replace generator with the generator from the ``FastForwardSampler``.
+                    dataset_dict[generator_attr_name] = iter(sampler)
+
+        # reset state dict.
+        self.state_dict = None
+
+    def reset_on_epoch(self) -> None:
+        self.state_dict = None
+
+    def __iter__(self) -> Iterator:
+        # create a generator from the wrapped Iterative Dataset
+        # if the dataset contained samplers, they will be transformers into generators
+        self.iter_data = iter(self.dataset)
+
+        # wrap any generator associated to a Sampler into a ``FastForwardSampler``.
+        self._wrap_generator_samplers()
+        return self
+
+    def __next__(self) -> Dict[str, Any]:
+        # fetch next data
+        data = next(self.iter_data)
+
+        # create current samplers state_dict
+        worker_info = get_worker_info()
+        state_dicts = {"id": worker_info.id if worker_info is not None else 0}
+        for k, v in self.samplers.items():
+            state_dicts.update({k: v.state_dict()})
+
+        # return both current data and samplers ``state_dict``.
+        return {"data": data, AutoRestartBatchKeys.PL_SAMPLERS: state_dicts}
+
+    @staticmethod
+    def convert_batch_into_state_dict(batch) -> Dict[str, Dict[int, Any]]:
+        """
+        This function is used to convert a batch into a state_dict
+        """
+        state_dict = {}
+        batch_worker_id = batch[AutoRestartBatchKeys.PL_SAMPLERS].pop("id")
+        worker_id = batch_worker_id[-1].item()
```

Comment:
Noob question: isn't `batch_worker_id` a single `int` here? what's the `[-1]` for?

Reply:
No, as it is actually a batch. So we just extract the last value.